Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+            {
+                doc.AbsoluteExpiration = options.AbsoluteExpiration.Value.UtcDateTime;
+            }
+            if (options.AbsoluteExpirationRelativeToNow.HasValue)
+            {
+                doc.AbsoluteExpiration = DateTime.UtcNow +
+                    options.AbsoluteExpirationRelativeToNow.Value;
+            }
+            return doc;
+        }
+
+        void IDistributedCache.Set(string key, byte[] value, DistributedCacheEntryOptions options) =>
+            _cacheEntries.Document(key).SetAsync(MakeCacheDoc(value, options)).Wait();
+
+        Task IDistributedCache.SetAsync(string key, byte[] value, DistributedCacheEntryOptions options,
+            CancellationToken token) =>
+            _cacheEntries.Document(key).SetAsync(MakeCacheDoc(value, options),
+            cancellationToken: token);
+
+        /// <summary>
+        /// Scans the firestore collection for expired cache entries and
+        /// deletes them.  
+        /// </summary>
+        /// <param name="token">A cancellation token.</param>
+        public async Task CollectGarbageAsync(CancellationToken token)
+        {
+            _logger.LogTrace("Begin garbage collection.");
+            // Purge entries whose AbsoluteExpiration has passed.
+            const int pageSize = 40;
+            int batchSize;  // Batches of cache entries to be deleted.
+            var now = DateTime.UtcNow;
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderByDescending("AbsoluteExpiration")
+                    .StartAfter(now)
+                    .Limit(pageSize)
+                    .GetSnapshotAsync(token);
+                batchSize = 0;
+                WriteBatch writeBatch = _cacheEntries.Database.StartBatch();
+                foreach (DocumentSnapshot docSnapshot in querySnapshot.Documents)
+                {
+                    if (docSnapshot.ConvertTo<CacheDoc>()
+                        .AbsoluteExpiration.HasValue)
+                    {
+                        writeBatch.Delete(docSnapshot.Reference,
+                            Precondition.LastUpdated(
+                                docSnapshot.UpdateTime.GetValueOrDefault()));
+                        batchSize += 1;
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);
+                    await writeBatch.CommitAsync(token);
+                }
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
+            } while (batchSize == pageSize);
+
+            // Purge entries whose SlidingExpiration has passed.
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderBy("LastRefresh")
+                    .Limit(pageSize)

```

Comment:
If we don't batch the deletes, presumably we don't need the limit or the loop? Just exit as soon as we find a `LastRefresh` that's older than we need?

Reply:
2 rpcs to delete each entry?  That sounds really slow.

Reply:
We can perform the RPCs in parallel, but I don't see that it would be two RPCs per entry - we'd just perform a single query and iterate over the results, firing off an RPC to delete each entry as we go, then waiting for all the RPCs to complete (or possibly even *not* waiting) at the end.